### ISSUE TITLE
"Task still has children after it has been executed" assertion failure when inside `task_group`

### ISSUE THREAD

**Issue by e4lam** [oneapi-src/oneTBB#707]
When running regular child task spawning inside a `task_group` in TBB 2020, it triggers the "Task still has children after it has been executed" assertion failure at [custom_scheduler.h:494](https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/src/tbb/custom_scheduler.h#L494),
```
__TBB_ASSERT( t->prefix().ref_count==0, "Task still has children after it has been executed" );
```

Normally, when all the children are done executing, the parent task's ref count will be reset to 0 at [custom_scheduler.h:793](https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/src/tbb/custom_scheduler.h#L793),
```
    if ( !ConcurrentWaitsEnabled(parent) ) {
        if ( parent.prefix().ref_count != 1) {
            // This is a worker that was revoked by the market.
            __TBB_ASSERT( worker_outermost_level(),
                "Worker thread exits nested dispatch loop prematurely" );
            return;
        }
        parent.prefix().ref_count = 0; <---- line 793
    }
```
However, when the parent task is being run inside a `task_group`, the `ConcurrentWaitsEnabled()` call here will return `true` instead of `false`, thus skipping the ref count reset. This in turn causes the assertion failure after the parent task finishes.

I was not able to track this down in the master branch since the code here has changed dramatically. I'm not able to upgrade from TBB 2020 because of the [VFX Platform](http://vfxplatform.com). Any advice or suggestions?

Here's a minimal reproducer using the FibTask example ([live at godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAMxmpK4AMngMmAByPgBGmMQgAKwAnKQADqgKhE4MHt6%2BAUEZWY4CYRHRLHEJKbaY9qUMQgRMxAR5Pn6BdQ05za0E5VGx8UmpCi1tHQXdEwNDldVjAJS2qF7EyOwc5v7hyN5YANQm/m4EMTEA9BcxAHQIp9gmGgCCu/uHmCdnThPEmFYTxe7zeByYCgURwAYngYgAVCEAayOICOaS8MVoeGQR1uIBALQUSJBJgA7FY3hisTiQCCjgyjmgGBMjuECEcWABPSKnSmvRlswQnMwANm5Qh8fNJb0Fc0cuPZRyUxDwYggSoYy3pjPJ/MFgrwVCOEAYPzcRzM2tlBttAIIGzNDGlNrtmAdxDNKrVtFNAFouMsRRZlfEff6rS6BbqyQARGXRhmwhHIjXChikIUc8yihQ%2Ba2Jg1o7mRU0F20MzMSnwQPMscsx/Ux%2BOuvGXAlElEAKkwqkwyC8BEwECDqAAbvFVVgdQy9TODUaTSXzUd/A3bXPWxXOVzJSwfrHQ6r1SXtf4mxu4/PBfUlNfG/fbUrVJmuVHt4KmIPUCLRUwD0cXYRAA7hAYi0KgohDgA%2BsgCB0OgI5BsmiLEhAy5%2BkcXCZqoZ4Xh%2BX5EL%2BMQAUBmCgeBkFMDBcEIUhMJwqhSLoTyRyYUERxcnhCYfoyEzoASWzEMQ5qnBa5iWNg0IAPIAErYGiEmiWcRwAlQsHrIII7KeJZhmCYiRuM6el8kcVxXEc6wEBiBBQhoj4GkoBDQWpGleFpa7vrxyppEwwEMBAMTcVuFYKL5/nQYY6DQcBTCENB/DEJFtC%2BkwwWFtu/GCZOOkimYrzQvC2CyWieW5a5aDuQQ2liWJeUGUZEmmeZlmDjZUJcA5grVvupyHqowacV5l4thljL2o6RwMF4KVpAQxDDScV5vOS8bngmY6oHg6BHOgqAAJLVQWm6FkqdaLYRP45p2pEgRA%2BKEsiBKUVBmAuagqBHchjGpv4mZ1ulgoPZ2BJhX5DDvZ9kUMNFsWEPdyLpatCZKiwcUBcdFLzi1BDwVCCgIKgwFQr2aT9kOO0aOimQEH6zL4I0qmYOplWCPOWUoDltUqU1/hrfzRwAOLSdJh59eLfMNcZ%2Bnra2e2HSOUbYxZuN4PjhPE3iCDfDEXjAEcwHawC1MTHTAgMzkTMs5pHJq1hQoTIC6DswQAmc8JuVNYZzqS77h4WK8Yt%2B%2BL3u8/hwPItBwDEOsaRHNHseLQnXhpHcxDuRA8sEIDjLJ6ncNHYtE2ekc9my%2B8y2gq8HCrLQnCJLwfgcFopCoJwbjWNYyrrJs3y7DwpAEJoNerEiSQaHcGiSKKiRkgAHJIkjJP4ZLJIv%2BicJIjfD63nC8AoIAaIPw%2BrHAsAwIgKCoCwaR0PE5CUGgN93wkwBcFwQQ0LQQ7EAfgU7zEcIrQuScAHk/NgghpIMFoCA5uvAsBoyMOIOBpB8AAgcHgCcB8UG9n7IObYLd2T1B3tiGIxBgEeCwDveaeAWCgJrnwAwwAFAADU8DkWkmTJuA9%2BCCBEGIdgUgZCCEUCodQKDdDYQMEYFAndLD6DhAfWAzA2CEkuKQCcCQOBmA0Do/wI9W5zRyNgv0/E%2BqmEsNYLgZI2L6z9LGeEFgLDQSEHIZQyh5JCCENBWM2APHYDcK8IqsZoIAFlsBeNeILCJ%2B96jkxyC4GG0w/DYVCOEYYVRRjYWKNkAQyS9A5MaAsEYCRsJ2HiQIfoUxPCdD0OUjBlTJiDHSYsLJtgmn5LKU04pmTSmrAUD3LYeh5qYG2DwWu9dt4oLbhwVQc9RR%2BlFJIeOyBcTvzuGYE0HdLHyKOLgQgJART%2BGwkcDwz96AiX7ssXgQ84HLFHgEfwdx/AvNeW8t5G8OBb1IHQxIR8m4txmfvQ%2Bx87mkDPpfKyNkH4QCfrfC5ehMAMxIHoXhwhRDiCEWi0RahqEAjGbwYC5C0j0ImRwBupAAW8BmdJNqg5LLGjmQspZKy1lcA2Vs6%2B8L4hHMDDck%2BqxtZMCwAkEcnzvm/P%2BTvIFtgQW3K0Pc0gY8l53GSIkfwkgzBcESH80UOjZ6JE%2Bfoyl0q96goVWSswvA6FcA0FK6ZZr5UGM0VkZwkggA%3D%3D%3D)),
```
#include <tbb/tbb.h>
#include <iostream>

class FibTask : public tbb::task
{
public:
    const int myN;
    int &mySum;

    static int serial(int n)
    {
        if (n < 2)
            return n;
        return serial(n-1) + serial(n-2);
    }

    FibTask(int n, int &sum)
        : myN(n)
        , mySum(sum)
    {
    }
    tbb::task *execute() override
    {
        if (myN < 3)
        {
            mySum = serial(myN);
        }
        else
        {
            int x, y;
            auto &a = *new(allocate_child()) FibTask(myN - 1, x);
            auto &b = *new(allocate_child()) FibTask(myN - 2, y);

            std::cerr << "BEFORE: " << ref_count() << "\n"; // outputs 0
            set_ref_count(3);
            spawn(b);
            spawn_and_wait_for_all(a);
            std::cerr << "AFTER:  " << ref_count() << "\n"; // outputs 1
            mySum = x + y;
        }
        return nullptr;
    }
};

void doIt()
{
    int sum;
    auto &task = *new(tbb::task::allocate_root()) FibTask(3, sum);
    tbb::task::spawn_root_and_wait(task);
}

int main()
{
    // this shows expected 0 post-condition ref_count
    std::cerr << "=== GOOD ===\n";
    doIt();

    // this shows the bug where post-condition ref_count is 1 instead
    std::cerr << "\n=== BAD ===\n";
    tbb::task_group group;
    group.run(doIt);
    group.wait();
    return 0;
}
```
It outputs:
```
=== GOOD ===
BEFORE: 0
AFTER:  0

=== BAD ===
BEFORE: 0
AFTER:  1
```

**Comment by anton-potapov** [comment]
@e4lam , see below :
Short answer is to use explicit `task_group_context` for root task allocated inside `doIt()`.

Details : 
1. each `task` has an associated `task_group_context` 
2. this context can be explicitly specified during root task [allocation ](https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/include/tbb/task.h#L669)
3. by default context of the current task is used 
4. `wait_for_all` checks `concurrent_wait` property of context associated with the task being waited on whether or not to reset root `ref_count` to 0 
5. each `task_group` has it own associated[`task_group_context`](https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/include/tbb/task_group.h#L105)
6. this context is created with `concurrent_wait` feature [enabled](https://github.com/oneapi-src/oneTBB/blob/eca91f16d7490a8abfdee652dadf457ec820cc37/include/tbb/task_group.h#L194)
7. So, when `doIt()` function in your sample allocates a root task, it's associated context is different in two cases :
    - default one when called inside `main()`, with `concurrent_wait` disabled
    - `task_group`'s one with `concurrent_wait` enabled 
8. Explicit `task_group_context` should be used to always disable `concurrent_wait`, as your code can guarantee this.     

Here is [slightly modified version](or [live](https://godbolt.org/z/d5938hKor) ) of the sample for more details : 
```c++
#include <tbb/tbb.h>
#include <iostream>

class FibTask : public tbb::task
{
public:
    const int myN;
    int &mySum;

    static int serial(int n)
    {
        if (n < 2)
            return n;
        return serial(n-1) + serial(n-2);
    }

    FibTask(int n, int &sum)
        : myN(n)
        , mySum(sum)
    {
    }
    tbb::task *execute() override
    {
        if (myN < 3)
        {
            mySum = serial(myN);
        }
        else
        {
            int x, y;
            auto &a = *new(allocate_child()) FibTask(myN - 1, x);
            auto &b = *new(allocate_child()) FibTask(myN - 2, y);

            std::cerr << "BEFORE: " << ref_count() << "\n"; // outputs 0
            set_ref_count(3);
            spawn(b);
            spawn_and_wait_for_all(a);
            std::cerr << "AFTER:  " << ref_count() << "\n"; // outputs 1
            mySum = x + y;
        }
        return nullptr;
    }
};

//uncomment this to fix the sample :)
//#define USE_EXPLICIT_CONTEXT 1

void doIt()
{
    int sum;

#if USE_EXPLICIT_CONTEXT
    tbb::task_group_context ctx;
    auto &task = *new(tbb::task::allocate_root(ctx)) FibTask(3, sum);
#else
    auto &task = *new(tbb::task::allocate_root()) FibTask(3, sum);
#endif    

    bool concurrent_wait_enabled = task.group()->traits() & tbb::task_group_context::concurrent_wait;
    std::cerr << "concurrent_wait enabled : " << concurrent_wait_enabled << std::endl;
    tbb::task::spawn_root_and_wait(task);
}

int main()
{
    // this shows expected 0 post-condition ref_count
    std::cerr << "=== GOOD ===\n";
    doIt();

    // this shows the bug where post-condition ref_count is 1 instead
    std::cerr << "\n=== BAD ===\n";
    tbb::task_group group;
    group.run(doIt);
    group.wait();
    return 0;
}


```

**Comment by e4lam** [comment]
@anton-potapov Thank you for the details! This is much more than I was able to figure out myself. A few questions though:

- What do we lose by turning off `concurrent_wait`? I'd presume that this is turned on for `task_group`'s for a reason? All I could find in the docs was that it "... controls the reference-counting behavior of methods task::wait_for_all and task::spawn_and_wait_for_all" but that's not really saying much.
- Specifying an explicit `task_group_context` for `allocate_root` seems to have its own drawbacks. If I understand correctly, doing so means that task group cancellation will no longer work unless I can ensure that all root tasks manually spawned within the group share the same `task_group_context`?
- Is there any real harm with the current situation? If not, I'm inclined to just comment out the assertion in my custom build. :)

Thanks for your insight!

**Comment by pavelkumbrasev** [comment]
Hi @e4lam, I'm not so proficient as @anton-potapov in old TBB and can not give you all answers.
After a quick check seems that it should be ok to just comment the assert. Have you already tested this approach?

**Comment by e4lam** [comment]
Hi @pavelkumbrasev I've long worked around this issue by using `tbb::parallel_invoke()` instead. At this point, no one is going back to TBB 2020 so I'll just close this issue as wontfix.
